### PR TITLE
[Backport] Backport JFR Streaming API to Dragonwell 8

### PR DESCRIPTION
**Pull request by D-D-H**
issue: https://github.com/alibaba/dragonwell8/issues/112

### CODE REVIEW

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp [hunk 1]
```diff
@@ -578,10 +600,153 @@ JfrJavaSupport::CAUSE JfrJavaSupport::cause() {
...
+  ThreadExclusionListAccess lock;
+  if (thread_is_not_excluded(h_obj)) {
+    return false;
+  }
+  remove_thread_from_exclusion_list(h_obj);
+  return true;
+}
+
+jlong JfrJavaSupport::jfr_thread_id(jobject thread) {
 //  ThreadsListHandle tlh;
```
Comment (kuaiwei) [inline src/share/vm/jfr/jni/jfrJavaSupport.cpp:697]: remove unused comments

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp [hunk 2]
```diff
@@ -56,6 +57,12 @@ void JfrJavaSupport::check_java_thread_in_native(Thread* t) {
   assert(t->is_Java_thread(), "invariant");
   assert(((JavaThread*)t)->thread_state() == _thread_in_native, "invariant");
 }
+
+static void check_new_unstarted_java_thread(Thread* t) {
+  assert(t != NULL, "invariant");
+  assert(t->is_Java_thread(), "invariant");
+  // assert(((JavaThread*)t)->thread_state() == _thread_new, "invariant");
```
Comment (kuaiwei) [inline src/share/vm/jfr/jni/jfrJavaSupport.cpp:64]: This line is in jdk master. Why comment this line? please add a comment here?

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp [hunk 3]
```diff
@@ -578,10 +600,153 @@ JfrJavaSupport::CAUSE JfrJavaSupport::cause() {
 //const char* const JDK_JFR_MODULE_NAME = "jdk.jfr";
```
Comment (kuaiwei) [inline src/share/vm/jfr/jni/jfrJavaSupport.cpp:600]: remove these line

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp [hunk 4]
```diff
@@ -578,10 +600,153 @@ JfrJavaSupport::CAUSE JfrJavaSupport::cause() {
...
+  }
+  remove_thread_from_exclusion_list(h_obj);
+  return true;
+}
+
+jlong JfrJavaSupport::jfr_thread_id(jobject thread) {
 //  ThreadsListHandle tlh;
   // XXX is it correct and safe?
-  JavaThread* native_thread = java_lang_Thread::thread(JNIHandles::resolve_non_null(target_thread));
+  JavaThread* native_thread = java_lang_Thread::thread(JNIHandles::resolve_non_null(thread));
```
Comment (kuaiwei) [inline src/share/vm/jfr/jni/jfrJavaSupport.cpp:699]: tlh is used to protect thread handle, do we have similar mechanism in jdk8?

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp [hunk 5]
```diff
@@ -578,10 +600,153 @@ JfrJavaSupport::CAUSE JfrJavaSupport::cause() {
...
 //  (void)tlh.cv_internal_thread_to_JavaThread(target_thread, &native_thread, NULL);
   return native_thread != NULL ? JFR_THREAD_ID(native_thread) : 0;
 }
+
+void JfrJavaSupport::exclude(jobject thread) {
+  HandleMark hm;
+  // ThreadsListHandle tlh;
+  // JavaThread* native_thread = NULL;
+  // (void)tlh.cv_internal_thread_to_JavaThread(thread, &native_thread, NULL);
+  JavaThread* native_thread = java_lang_Thread::thread(JNIHandles::resolve_non_null(thread));
```
Comment (kuaiwei) [inline src/share/vm/jfr/jni/jfrJavaSupport.cpp:709]: In jdk master, this is wrapped in get_native, I think we can follow to use it

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp [hunk 6]
```diff
@@ -578,10 +600,153 @@ JfrJavaSupport::CAUSE JfrJavaSupport::cause() {
...
+    add_thread_to_exclusion_list(thread);
+  }
+}
+
+void JfrJavaSupport::include(jobject thread) {
+  HandleMark hm;
+  // ThreadsListHandle tlh;
+  // JavaThread* native_thread = NULL;
+  // (void)tlh.cv_internal_thread_to_JavaThread(thread, &native_thread, NULL);
+  JavaThread* native_thread = java_lang_Thread::thread(JNIHandles::resolve_non_null(thread));
```
Comment (kuaiwei) [inline src/share/vm/jfr/jni/jfrJavaSupport.cpp:723]: use get_native

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp [hunk 7]
```diff
@@ -578,10 +600,153 @@ JfrJavaSupport::CAUSE JfrJavaSupport::cause() {
...
+    remove_thread_from_exclusion_list(thread);
+  }
+}
+
+bool JfrJavaSupport::is_excluded(jobject thread) {
+  HandleMark hm;
+  // ThreadsListHandle tlh;
+  // JavaThread* native_thread = NULL;
+  // (void)tlh.cv_internal_thread_to_JavaThread(thread, &native_thread, NULL);
+  JavaThread* native_thread = java_lang_Thread::thread(JNIHandles::resolve_non_null(thread));
```
Comment (kuaiwei) [inline src/share/vm/jfr/jni/jfrJavaSupport.cpp:737]: use get_native

File: src/share/vm/jfr/jni/jfrJniMethod.cpp [hunk 8]
```diff
@@ -303,6 +310,10 @@ JVM_ENTRY_NO_ENV(jboolean, jfr_event_writer_flush(JNIEnv* env, jclass cls, jobje
   return JfrJavaEventWriter::flush(writer, used_size, requested_size, thread);
 JVM_END
 
+JVM_ENTRY_NO_ENV(void, jfr_flush(JNIEnv* env, jobject jvm))
```
Comment (kuaiwei) [inline src/share/vm/jfr/jni/jfrJniMethod.cpp:313]: I'm wondering why jfr entry always has a jvm argument and not use it.

File: src/share/vm/jfr/leakprofiler/chains/edgeStore.cpp [hunk 9]
```diff
@@ -75,22 +80,21 @@ bool EdgeStore::contains(const oop* reference) const {
 
 StoredEdge* EdgeStore::get(const oop* reference) const {
   assert(reference != NULL, "invariant");
-  const StoredEdge e(NULL, reference);
-  EdgeEntry* const entry = _edges->lookup_only(e, (uintptr_t)reference);
+  EdgeEntry* const entry = _edges->lookup_only((uintptr_t)reference);
   return entry != NULL ? entry->literal_addr() : NULL;
 }
 
 StoredEdge* EdgeStore::put(const oop* reference) {
```
Comment (kuaiwei) [inline src/share/vm/jfr/leakprofiler/chains/edgeStore.cpp:87]: Does gc scan EdgeStore as root set?

File: src/share/vm/jfr/jni/jfrJniMethod.cpp (hunk 8 above)
Comment (D-D-H) [reply src/share/vm/jfr/jni/jfrJniMethod.cpp:313]: it isn't a static method and the argument 'jvm' is the receiver

File: src/share/vm/jfr/leakprofiler/chains/edgeStore.cpp (hunk 9 above)
Comment (D-D-H) [reply src/share/vm/jfr/leakprofiler/chains/edgeStore.cpp:87]: this operation is invoked  at safepoint.

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp (hunk 5 above)
Comment (D-D-H) [reply src/share/vm/jfr/jni/jfrJavaSupport.cpp:709]: get_native was introduced in http://hg.openjdk.java.net/jdk/jdk/rev/ac646e22b4df, this patch is not included in backport list

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp (hunk 1 above)
Comment (D-D-H) [reply src/share/vm/jfr/jni/jfrJavaSupport.cpp]: fixed

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp (hunk 2 above)
Comment (D-D-H) [reply src/share/vm/jfr/jni/jfrJavaSupport.cpp:66]: fixed

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp (hunk 4 above)
Comment (D-D-H) [reply src/share/vm/jfr/jni/jfrJavaSupport.cpp:699]: no, openjdk8 didn't support thread smr

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp (hunk 3 above)
Comment (D-D-H) [reply src/share/vm/jfr/jni/jfrJavaSupport.cpp]: fixed

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp (hunk 6 above)
Comment (D-D-H) [reply src/share/vm/jfr/jni/jfrJavaSupport.cpp:712]: same as above

File: src/share/vm/jfr/jni/jfrJavaSupport.cpp (hunk 7 above)
Comment (D-D-H) [reply src/share/vm/jfr/jni/jfrJavaSupport.cpp:723]: same as above

File: src/share/vm/jfr/recorder/checkpoint/types/jfrType.cpp [hunk 10]
```diff
@@ -88,27 +87,18 @@ class JfrCheckpointThreadClosure : public ThreadClosure {
...
-  if (tl->is_dead()) {
-    return;
-  }
   ++_count;
-  _writer.write_key(tl->thread_id());
-  _writer.write(t->name());
-  const OSThread* const os_thread = t->osthread();
-  _writer.write<traceid>(os_thread != NULL ? os_thread->thread_id() : 0);
+  _writer.write_key(JfrThreadId::jfr_id(t));
+  const char* const name = JfrThreadName::name(t);
```
Comment (kuaiwei) [inline src/share/vm/jfr/recorder/checkpoint/types/jfrType.cpp:94]: ResourceMark is required for JfrThreadName::name

File: src/share/vm/jfr/recorder/checkpoint/types/jfrType.cpp [hunk 11]
```diff
@@ -293,59 +288,28 @@ void VMOperationTypeConstant::serialize(JfrCheckpointWriter& writer) {
...
 void ThreadStateConstant::serialize(JfrCheckpointWriter& writer) {
   JfrThreadState::serialize(writer);
 }
 
 void JfrThreadConstant::serialize(JfrCheckpointWriter& writer) {
   assert(_thread != NULL, "invariant");
   assert(_thread == Thread::current(), "invariant");
-  assert(_thread->is_Java_thread(), "invariant");
-  assert(!_thread->jfr_thread_local()->has_thread_checkpoint(), "invariant");
-  ResourceMark rm(_thread);
```
Comment (kuaiwei) [inline src/share/vm/jfr/recorder/checkpoint/types/jfrType.cpp:337]: ResourceMark is removed. But thread::name need a mark. I think ResourceMark should be declared in other place, please add a comment to point it.

File: src/share/vm/jfr/recorder/checkpoint/types/jfrType.cpp (hunk 10 above)
Comment (D-D-H) [reply src/share/vm/jfr/recorder/checkpoint/types/jfrType.cpp:94]: ResourceMark is created at JfrCheckpointManager::write_threads

File: src/share/vm/jfr/recorder/checkpoint/types/jfrType.cpp (hunk 11 above)
Comment (D-D-H) [reply src/share/vm/jfr/recorder/checkpoint/types/jfrType.cpp:337]: same as above

**Pull request merged**